src/test/java/htsjdk/samtools/cram/compression/rans/RansTest.java

```diff
@@ -49,107 +57,145 @@ public String toString() {
+        final ByteBuffer compressed = ransEncode.compress(data, params);
+        // first byte of compressed data gives the order
+        Assert.assertEquals(compressed.get(), (byte) params.getOrder().ordinal());
+        // the next 4 bytes gives the compressed size
         Assert.assertEquals(compressed.getInt(), compressed.limit() - 9);
-        Assert.assertEquals(compressed.getInt(), data.length);
-    }
-
-    private byte[] getNBytesWithValues(final int n, final BiFunction<Integer, Integer, Byte> valueForIndex) {
-        final byte[] data = new byte[n];
-        for (int i = 0; i < data.length; i++) {
-            data[i] = valueForIndex.apply(n, i);
-        }
-        return data;
+        // the next 4 bytes gives the uncompressed size
+        Assert.assertEquals(compressed.getInt(), data.limit());
     }
 
-    private static void roundTripForEachOrder(final ByteBuffer data) {
-        for (RANS.ORDER order : RANS.ORDER.values()) {
-            roundTripForOrder(data, order);
-            data.rewind();
+    @Test(dataProvider = "ransNx16")
+    public void testRansNx16Header(
+            final RANSNx16Encode ransEncode,
+            final RANSNx16Decode unused,
+            final RANSNx16Params params) {
+        final int size = 1000;
+        final ByteBuffer data = ByteBuffer.wrap(randomBytesFromGeometricDistribution(size, 0.01));
+        final ByteBuffer compressed = ransEncode.compress(data, params);
+        // first byte of compressed data gives the formatFlags
+        Assert.assertEquals(compressed.get(), (byte) params.getFormatFlags());
+        // if nosz flag is not set, then the uncompressed size is recorded
+        if (!params.getNosz()){
+            Assert.assertEquals(Utils.readUint7(compressed), size);
         }
     }
 
-    private static void roundTripForEachOrder(final byte[] data) {
-        for (RANS.ORDER order : RANS.ORDER.values()) {
-            roundTripForOrder(data, order);
-        }
+    @Test(dataProvider="allRansAndData")
+    public void testRANS(
+            final TestCaseWrapper tc,
+            final RANSEncode ransEncode,
+            final RANSDecode ransDecode,
+            final RANSParams params) {
+        ransRoundTrip(ByteBuffer.wrap(tc.testArray), ransEncode, ransDecode, params);
     }
 
-    private static void roundTripForOrder(final ByteBuffer data, final RANS.ORDER order) {
-        final RANS rans = new RANS();
-        final ByteBuffer compressed = rans.compress(data, order);
-        final ByteBuffer uncompressed = rans.uncompress(compressed);
+    private static void ransRoundTrip(
+            final ByteBuffer data,
+            final RANSEncode ransEncode,
+            final RANSDecode ransDecode,
+            final RANSParams params) {
+        final ByteBuffer compressed = ransEncode.compress(data, params);
+        final ByteBuffer uncompressed = ransDecode.uncompress(compressed, params);
```

Comment:
Again, I think you didn't write this, but isn't the while loop below just a long winded way of saying of `Bytebuffer.equals`. Probably have to check  the `ByteBuffer` contract is, but it seems like it would be much less verbose.